File: Examples/monitor.js
#!/usr/bin/env node
/**
 * Water Sensor Monitor - Node.js Integration Example
 * Polls the Arduino water sensor API and provides notifications
 */

const axios = require('axios');
const fs = require('fs');

// Configuration
const ARDUINO_IP = '192.168.1.100'; // Change to your Arduino's IP
const API_URL = `http://${ARDUINO_IP}/api/status`;
const POLL_INTERVAL = 5000; // milliseconds
const LOG_FILE = 'sensor_log.json';

class WaterSensorMonitor {
    constructor(arduinoIp) {
        this.apiUrl = `http://${arduinoIp}/api/status`;
        this.lastAlertCount = 0;
        this.consecutiveFailures = 0;
        this.maxFailures = 5;
        this.logData = [];
    }

    /**
     * Fetch current status from Arduino API
     */
    async fetchStatus() {
        try {
            const response = await axios.get(this.apiUrl, { timeout: 5000 });
            this.consecutiveFailures = 0;
            return response.data;
        } catch (error) {
            this.consecutiveFailures++;
            
            if (error.code === 'ECONNABORTED') {
                console.log('⚠️  Connection timeout - Arduino may be offline');
            } else if (error.code === 'ECONNREFUSED') {
                console.log('⚠️  Connection refused - Check Arduino IP and port');
            } else {
                console.log(`⚠️  Error: ${error.message}`);
            }
            
            return null;
        }
    }

    /**
     * Log sensor data to JSON file
     */
    logReading(data) {
        if (!data) return;

        const logEntry = {
            timestamp: new Date().toISOString(),
            ...data
        };

        this.logData.push(logEntry);

        // Write to file every 10 readings to reduce I/O
        if (this.logData.length % 10 === 0) {
            this.saveLog();
        }
    }

    /**
     * Save log data to file
     */
    saveLog() {
        try {
            fs.writeFileSync(LOG_FILE, JSON.stringify(this.logData, null, 2));
        } catch (error) {
            console.log(`⚠️  Error saving log: ${error.message}`);
        }
    }

    /**
     * Check if a new alert has occurred
     */
    checkNewAlert(data) {
        if (!data) return false;

        const currentAlerts = data.totalAlerts;
        if (currentAlerts > this.lastAlertCount) {
            this.lastAlertCount = currentAlerts;
            return true;
        }
        return false;
    }

    /**
     * Send notification for new alert
     */
    sendNotification(data) {
        console.log('\n' + '='.repeat(50));
        console.log('🚨 NEW WATER ALERT!');
        console.log('='.repeat(50));
        console.log(`Time: ${new Date().toLocaleString()}`);
        console.log(`Sensor Value: ${data.sensorValue}`);
        console.log(`Total Alerts: ${data.totalAlerts}`);
        console.log('='.repeat(50) + '\n');

        // Add your notification code here:
        // - Send email
        // - Send SMS
        // - Push notification
        // - Webhook
    }

    /**
     * Display current status to console
     */
    displayStatus(data) {
        if (!data) return;

        const status = data.waterDetected ? '🚨 WATER DETECTED' : '✅ All Clear';
        const value = data.sensorValue.toString().padStart(4);
        const alerts = data.totalAlerts.toString().padStart(3);
        const uptime = data.uptime.padStart(12);
        const signal = data.signalStrength.toString().padStart(4);

        console.log(
            `${status} | ` +
            `Value: ${value} | ` +
            `Alerts: ${alerts} | ` +
            `Uptime: ${uptime} | ` +
            `Signal: ${signal} dBm`
        );
    }

    /**
     * Check connection health
     */
    checkHealth() {
        if (this.consecutiveFailures >= this.maxFailures) {
            console.log(`\n⚠️  WARNING: ${this.consecutiveFailures} consecutive failures!`);
            console.log('Arduino may be offline or unreachable');
            console.log('Check power, WiFi, and network connectivity\n');
            return false;
        }
        return true;
    }

    /**
     * Main monitoring loop
     */
    async run(pollInterval = 5000) {
        console.log('='.repeat(60));
        console.log('Water Sensor Monitor - Node.js Integration');
        console.log('='.repeat(60));
        console.log(`Arduino IP: ${ARDUINO_IP}`);
        console.log(`Polling interval: ${pollInterval / 1000} seconds`);
        console.log(`Log file: ${LOG_FILE}`);
        console.log('='.repeat(60));
        console.log('\nPress Ctrl+C to stop\n');

        // Main loop
        const intervalId = setInterval(async () => {
            // Fetch current status
            const data = await this.fetchStatus();

            // Check connection health
            if (!this.checkHealth()) {
                return;
            }

            // Display status
            this.displayStatus(data);

            // Log data
            this.logReading(data);

            // Check for new alerts
            if (this.checkNewAlert(data)) {
                this.sendNotification(data);
            }
        }, pollInterval);

        // Handle graceful shutdown
        process.on('SIGINT', () => {
            clearInterval(intervalId);
            this.saveLog();
            console.log('\n\nMonitoring stopped by user');
            console.log(`Log saved to: ${LOG_FILE}`);
            process.exit(0);
        });
    }
}

// Main entry point
async function main() {
    // Check if custom IP provided
    const arduinoIp = process.argv[2] || ARDUINO_IP;

    // Create monitor and run
    const monitor = new WaterSensorMonitor(arduinoIp);
    await monitor.run(POLL_INTERVAL);
}

// Run if called directly
if (require.main === module) {
    main().catch(error => {
        console.error('Fatal error:', error);
        process.exit(1);
    });
}

module.exports = WaterSensorMonitor;

File: Examples/monitor.py
#!/usr/bin/env python3
"""
Water Sensor Monitor - Python Integration Example
Polls the Arduino water sensor API and logs data
"""

import requests
import time
import json
import csv
from datetime import datetime
import sys

# Configuration
ARDUINO_IP = "192.168.1.100"  # Change to your Arduino's IP
API_URL = f"http://{ARDUINO_IP}/api/status"
POLL_INTERVAL = 5  # seconds
LOG_FILE = "sensor_log.csv"
ALERT_ENABLED = True


class WaterSensorMonitor:
    """Monitor Arduino water sensor via API"""
    
    def __init__(self, arduino_ip):
        self.api_url = f"http://{arduino_ip}/api/status"
        self.last_alert_count = 0
        self.consecutive_failures = 0
        self.max_failures = 5
        
        # Initialize CSV log file
        self.init_log_file()
    
    def init_log_file(self):
        """Create CSV log file with headers if it doesn't exist"""
        try:
            with open(LOG_FILE, 'x', newline='') as f:
                writer = csv.writer(f)
                writer.writerow([
                    'Timestamp',
                    'Water Detected',
                    'Sensor Value',
                    'Alert Active',
                    'Total Alerts',
                    'Last Alert',
                    'Uptime',
                    'Signal Strength'
                ])
            print(f"Created log file: {LOG_FILE}")
        except FileExistsError:
            print(f"Using existing log file: {LOG_FILE}")
    
    def fetch_status(self):
        """Fetch current status from Arduino API"""
        try:
            response = requests.get(self.api_url, timeout=5)
            response.raise_for_status()
            
            data = response.json()
            self.consecutive_failures = 0
            return data
            
        except requests.exceptions.Timeout:
            print("⚠️  Connection timeout - Arduino may be offline")
            self.consecutive_failures += 1
            return None
            
        except requests.exceptions.ConnectionError:
            print("⚠️  Connection error - Check network and Arduino")
            self.consecutive_failures += 1
            return None
            
        except requests.exceptions.RequestException as e:
            print(f"⚠️  Request error: {e}")
            self.consecutive_failures += 1
            return None
            
        except json.JSONDecodeError:
            print("⚠️  Invalid JSON response")
            self.consecutive_failures += 1
            return None
    
    def log_data(self, data):
        """Log sensor data to CSV file"""
        if data is None:
            return
        
        try:
            with open(LOG_FILE, 'a', newline='') as f:
                writer = csv.writer(f)
                writer.writerow([
                    datetime.now().isoformat(),
                    data['waterDetected'],
                    data['sensorValue'],
                    data['alertActive'],
                    data['totalAlerts'],
                    data['lastAlert'],
                    data['uptime'],
                    data['signalStrength']
                ])
        except Exception as e:
            print(f"⚠️  Logging error: {e}")
    
    def check_new_alert(self, data):
        """Check if a new alert has occurred"""
        if data is None:
            return False
        
        current_alerts = data['totalAlerts']
        if current_alerts > self.last_alert_count:
            self.last_alert_count = current_alerts
            return True
        return False
    
    def send_notification(self, data):
        """Send notification for new alert (placeholder)"""
        print("\n" + "=" * 50)
        print("🚨 NEW WATER ALERT!")
        print("=" * 50)
        print(f"Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print(f"Sensor Value: {data['sensorValue']}")
        print(f"Total Alerts: {data['totalAlerts']}")
        print("=" * 50 + "\n")
        
        # Add your notification code here:
        # - Send email
        # - Send SMS via Twilio
        # - Push notification
        # - Webhook to other service
    
    def display_status(self, data):
        """Display current status to console"""
        if data is None:
            return
        
        # Status symbol
        status = "🚨 WATER DETECTED" if data['waterDetected'] else "✅ All Clear"
        
        # Build status line
        print(f"{status} | "
              f"Value: {data['sensorValue']:4d} | "
              f"Alerts: {data['totalAlerts']:3d} | "
              f"Uptime: {data['uptime']:>12s} | "
              f"Signal: {data['signalStrength']:4d} dBm")
    
    def check_health(self):
        """Check if connection to Arduino is healthy"""
        if self.consecutive_failures >= self.max_failures:
            print(f"\n⚠️  WARNING: {self.consecutive_failures} consecutive failures!")
            print("Arduino may be offline or unreachable")
            print("Check power, WiFi, and network connectivity\n")
            return False
        return True
    
    def run(self, poll_interval=5):
        """Main monitoring loop"""
        print("=" * 60)
        print("Water Sensor Monitor - Python Integration")
        print("=" * 60)
        print(f"Arduino IP: {ARDUINO_IP}")
        print(f"Polling interval: {poll_interval} seconds")
        print(f"Log file: {LOG_FILE}")
        print("=" * 60)
        print("\nPress Ctrl+C to stop\n")
        
        try:
            while True:
                # Fetch current status
                data = self.fetch_status()
                
                # Check connection health
                if not self.check_health():
                    time.sleep(poll_interval * 2)  # Wait longer on failures
                    continue
                
                # Display status
                self.display_status(data)
                
                # Log data
                self.log_data(data)
                
                # Check for new alerts
                if ALERT_ENABLED and self.check_new_alert(data):
                    self.send_notification(data)
                
                # Wait before next poll
                time.sleep(poll_interval)
                
        except KeyboardInterrupt:
            print("\n\nMonitoring stopped by user")
            print(f"Log saved to: {LOG_FILE}")
            sys.exit(0)


def main():
    """Main entry point"""
    # Check if custom IP provided
    if len(sys.argv) > 1:
        arduino_ip = sys.argv[1]
    else:
        arduino_ip = ARDUINO_IP
    
    # Create monitor and run
    monitor = WaterSensorMonitor(arduino_ip)
    monitor.run(poll_interval=POLL_INTERVAL)


if __name__ == "__main__":
    main()
